### ISSUE TITLE
A Generic Subject Classification Variable is needed for filtering Ontologies Conveniently 

### ISSUE THREAD

**Issue by giraygi** [EBISPOT/OLS#514]
Filtering by DFG or Other Subject Classifications in the search and ontologies pages might be more convenient for users. I recommend a Generic Subject Classification variable that enables the configurator to label an ontology with a number of classifications. By using a flexible data structure, only **classification** variable will be defined in the backend.  The **classification** variable will be configured with a flexible list of entries as shown below:

**classification:** 
  - DFG: 
    - 4.4.1 
    - 3.2.5 
  - BK: 
    - test 
    - test2 
  - collection: 
    - NFDI4ING 
    - NFDI4CULTURE

or below (slightly different):

  **classification:** 
  - {DFG2: ["4.4", "3.2"], DFG3: ["4.4.1", "3.2.5"]}
  - {BK: [test2,test3], bk: [test,test4]}
  - {subject: ["Information Science","Management Information Systems"],subfield: ["System Analysis","System Design"}

We have a BETA stage solution for this  in TIB as you can see from the links below:

- https://service.tib.eu/ts4tib/ontologies
- https://service.tib.eu/ts4tib/search?q=
- https://service.tib.eu/ts4tib/ontologies/bot
- https://service.tib.eu/ts4tib/swagger-ui.html#/ontology-controller/getAvailableSchemaKeysUsingGET
- https://service.tib.eu/ts4tib/swagger-ui.html#/ontology-controller/getClassificationsForSchemasUsingGET
- https://service.tib.eu/ts4tib/swagger-ui.html#/ontology-controller/filterOntologiesByClassificationUsingGET

I didn't finalize this feature yet and it is still under development but let me know if it looks interesting to you.

Best,
Giray

**Comment by henrietteharmse** [comment]
The short answer is yes we are definitely interested in something like this but we intended to roll it out in the new version of OLS rather than the current version.

Here is the longer answer:
Adding categories to OLS has been under consideration for some time. See for example [ticket 43](https://github.com/EBISPOT/OLS/issues/43) ticket dating from 2016. 

So why has this not been implemented OLS as yet? There are at least 3 reasons that contributed to this:

1. Which categories should be used is unclear.  [ticket 43](https://github.com/EBISPOT/OLS/issues/43) suggest the use of EDAM. The OBO Foundry are using domains which I think can be arbitrary and is completely up to the discretion of the ontology creators. See `domain: ` fields in [OBO Config](https://github.com/OBOFoundry/OBOFoundry.github.io/blob/master/_config.yml). And there is your suggestion. 
2. Whatever the chosen solution is, it should be generic enough to serve the biological community in general. In particular this solution should be workable for the 260+ ontologies on EBI OLS instance. I think therefore it is critical to define this solution as a community effort and in particular in collaboration with the OBO Foundry.
3. A big challenge we have with OLS is the performance and scalability of indexing. A change like this will cause about 170 ontologies to be reindexed on EBI OLS. Reindexing that many ontologies is likely to take about 2-3 weeks to reindex assuming there are no unforeseen problems. This is a painful and errorprone process. Moreover, no other new changes to ontologies will be able to process in this time. We therefore have to consider this change carefully to ensure that is will not adversely affect EBI OLS. This is why we thought it best to rather include it as part of the redesign of OLS. 

I think in order to move forward with this, we can do the following:

1. I will try to get the input others wrt this feature.
2. Create a PR with the changes you made so that we can get an idea of the impact of the changes.
3. Please test your changes using the minimal set of ontologies we use to test with. See [minimal test config](https://github.com/EBISPOT/OLS-Config/tree/minimal_test/ols-config-importer). 
4. We will test the minimal config and more in our environment to get better feel for the performance implications.
5. We will create a plan that will make this change available to the 170 over a period of say 3 months. 
6. Once 1-5 is complete/successful, I think we will be happy to make this change available in the existing version of OLS. 

@udp @matentzn @mcourtot @cmungall, if you have any thoughts on this, please feel free to comment.

**Comment by cthoyt** [comment]
Relevant discussion about how OBO Foundry would classify its ontologies on https://github.com/OBOFoundry/OBOFoundry.github.io/issues/1225

**Comment by udp** [comment]
I only have some specific feedback to add:

Regarding the UI:

* First I navigated to the ontologies page https://service.tib.eu/ts4tib/ontologies, but it seems there are no links to the classifications there?
* I then tried the search https://service.tib.eu/ts4tib/search?q=. I can select collections and schemas from the dropdowns on the left, but then on the page to which it navigates, my selection in the dropdown is gone.

More generally:

I am rather confused about the collections, classifications, and schemas. Are these all the same thing? From the API the "schema" seems to be a simple key value map per ontology. In the config, it seems only classifications are added. An overview of what these three concepts are and how they fit together would be very helpful.

**Comment by udp** [comment]
Ok I now understand from the description in #518 that there is a key value map, where "schema" refers to the keys and "classifications" refers to the values. The UI also makes more sense now, in that once you select a schema (key), the classifications dropdown appears (though neither picker seems to actually change the selected option in the dropdown, it does filter the results).

For the specific case of the "collection" key, you give its own dropdown. Can we do this for all keys instead in the UI, making the keys into subheadings? Then (a) there would no longer be a need to special case "collection" and (b) we could drop the "schema/classification" terminology from the UI, which I think is confusing as they are already overloaded terms in the context of ontologies.

**Comment by udp** [comment]
<img width="569" alt="Screenshot 2021-11-23 at 13 48 58" src="https://user-images.githubusercontent.com/454895/143036015-328e1b2f-ac85-4975-8c0f-f6db61a785f4.png">

**Comment by giraygi** [comment]
@henrietteharmse 

Thank you for your detailed reply. I would like to briefly summarize you my understanding of generic classification with toy examples in the obo-config.yaml.txt and ols-config.yaml.txt files in the attachment. According to this approach, the configurator is able to introduce new schema keys simply by defining them under the classification variable in the config files. So it is not needed to hardcode new schemas into OLS and ontologies can be classified according to new schemas. It will be very good if you have alternative schemas in mind and put them in the test dataset you have mentioned earlier.
[obo-config.yaml.txt](https://github.com/EBISPOT/OLS/files/7588403/obo-config.yaml.txt)
[ols-config.yaml.txt](https://github.com/EBISPOT/OLS/files/7588404/ols-config.yaml.txt)
o

**Comment by udp** [comment]
<img width="1034" alt="Screenshot 2021-11-23 at 13 52 46" src="https://user-images.githubusercontent.com/454895/143036658-29fa16da-3fe1-4d0a-9013-25e0641d1850.png">

**Comment by giraygi** [comment]
> <img alt="Screenshot 2021-11-23 at 13 48 58" width="569" src="https://user-images.githubusercontent.com/454895/143036015-328e1b2f-ac85-4975-8c0f-f6db61a785f4.png">

@udp I did not propose the first "Collection" dropdown of the TIB TS to EBI because the variable "collection" is a very special one to us. EBI does not need that. I only put Schemas and Collections for your convenience because that gives you the flexibility to introduce as many schemas as you wish. Collection is just a single schema.

**Comment by giraygi** [comment]
@udp This is actually how my pull request looks with the toy example I mentioned earlier.

![Screenshot from 2021-11-23 15-16-19](https://user-images.githubusercontent.com/48091473/143040688-c4a555ca-6111-46aa-9b0e-214bef14a8a5.png)

**Comment by udp** [comment]
Thanks. My point is that the UI would be easier to use if the schemes were the headings and the values were dropdowns, instead of separating schemas and classifications. I think it is quite confusing at the moment.

**Comment by giraygi** [comment]
> Thanks. My point is that the UI would be easier to use if the schemes were the headings and the values were dropdowns, instead of separating schemas and classifications. I think it is quite confusing at the moment.

Hello. I changed the frontend as you described. Now, the dropdown of each schema is created dynamically in a loop. One small note is that the dropdowns won't appear if you don't have any schema in your system. In addition, the previous ontology selections will be reset when you select a classification value from a particular schema. I did this to simplify filtering scenarios. On the contrary, term selections will persist. Below you see an instance with toy data. Please keep in mind that the number of schemas is a little bit exaggerated on the toy data.
![Screenshot from 2021-11-25 10-52-44](https://user-images.githubusercontent.com/48091473/143426298-4af92a4f-2872-49d0-a719-13ae80f4bfb0.png)

**Comment by udp** [comment]
This is much better, thank you for making the change! I will have a look at the code tomorrow and review.

**Comment by giraygi** [comment]
> This is much better, thank you for making the change! I will have a look at the code tomorrow and review.

I came across a small bug in the last minute and changed the pull request status to draft. Classification values that include a comma are automatically split as different request parameters now. I had solved this in the controller in the previous UI with a new InitBinder behaviour but now the current solution is not controller based. I will solve this tomorrow.
